src/sentry/tasks/data_export.py

```diff
@@ -1,8 +1,159 @@
+        "last_seen": item.last_seen.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+        "first_seen": item.first_seen.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+    }
+    if key == "user":
+        euser = item._eventuser
+        result["id"] = euser.ident if euser else ""
+        result["email"] = euser.email if euser else ""
+        result["username"] = euser.username if euser else ""
+        result["ip_address"] = euser.ip_address if euser else ""
+    return result
+
 
 @instrumented_task(name="sentry.tasks.data_export.compile_data", queue="data_export")
 def compile_data(data_export):
-    # TODO(Leander): Implement logic to compile CSV data via provided ExportedData object
+    if data_export.query_type == ExportQueryType.DISCOVER_V2:
+        process_discover_v2(data_export)
+        return
+    elif data_export.query_type == ExportQueryType.BILLING_REPORT:
+        process_billing_report(data_export)
+        return
+    elif data_export.query_type == ExportQueryType.ISSUE_BY_TAG:
+        file_path, file_name = process_issue_by_tag(data_export)
+        store_file(data_export, file_path, file_name)
+
+
+def process_discover_v2(data_export):
+    # TODO(Leander): Implement processing for Discover V2
+    return
+
+
+def process_billing_report(data_export):
+    # TODO(Leander): Implement processing for Billing Reports
+    return
+
+
+def process_issue_by_tag(data_export):
+    """
+    Convert tag payload to a CSV, returns (file_path, file_name) as a tuple
+    (Adapted from 'src/sentry/web/frontend/group_tag_export.py')
+    """
+    # Get the pertaining project
+    payload = data_export.query_info
+    project = Project.objects.get(slug=payload["project_slug"])
+
+    # Get the pertaining issue
+    group, _ = get_group_with_redirect(
+        payload["group_id"], queryset=Group.objects.filter(project=project)
+    )
+
+    # Get the pertaining key
+    key = payload["key"]
+    lookup_key = u"sentry:{0}".format(key) if tagstore.is_reserved_key(key) else key
+
+    # If the key is the 'user' tag, attach the event user
+    def attach_eventuser(items):
+        users = EventUser.for_tags(group.project_id, [i.value for i in items])
+        for item in items:
+            item._eventuser = users.get(item.value)
+
+    # Create the fields/callback lists
+    if key == "user":
+        callbacks = [attach_eventuser]
+        fields = [
+            "value",
+            "id",
+            "email",
+            "username",
+            "ip_address",
+            "times_seen",
+            "last_seen",
+            "first_seen",
+        ]
+    else:
+        callbacks = []
+        fields = ["value", "times_seen", "last_seen", "first_seen"]
+
+    iteration = 0
+    # Example file name: ISSUE_BY_TAG-project10-user__721.csv
+    file_details = u"{}-{}__{}".format(payload["project_slug"], key, data_export.id)
+    file_name = get_file_name(ExportQueryType.ISSUE_BY_TAG_STR, file_details)
+    file_path = path.join(tmp_dir, file_name)
+
```

Comment:
We could just use `tempfile.NamedTemporaryFile` here, and only ever open the file once. Then we can also use it as a context manager, and so when we're done it'll be cleaned up automatically. Same for the dict writer, we should probably just instantiate it out here, write the header and just keep writing to it. That way we avoid passing the `include_header` param too.

Reply:
That's actually exactly what I thought too when I saw `NamedTemporaryFile` in the docs, but I was trying to follow what was done in `src/sentry/tasks/assemble.py` 

I'll refactor it!